IRL

muizzzafar11 opened the pull request:
Matches user names and assigns them roles if the name matches the name in the given list. 

All feedback is appreciated. I tried to make the code as DRY as possible but I still feel like I have too much unnecessary stuff going on in the algorithm, or maybe too little? Nevertheless, let me know if there's anything more to add or any queries that can be further optimized.

muizzzafar11 committed:
added nameMatch and asigning role algorithm

mk5med commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+
+                    int spaceIndex = fullName.LastIndexOf(' ');
+                    // If a user doesn't have a first or last name in the server
+                    if (spaceIndex <= 3) {
```
This will skip people with first names that are two letters long

muizzzafar11 replied:
The idea was that people who only have two chars in their name [no first or last name] for example:
![image](https://user-images.githubusercontent.com/43075405/91503280-126c9380-e898-11ea-82cd-429fa7f3770a.png)
The algorithm will check if those two characters are found in the list given and send the message to inform the user.

muizzzafar11 replied:
I think there are only 1 or 2 people with server names 2 char long so I'll take this out since it won't be that useful.

muizzzafar11 replied:
I didn't think that there would be people with first name of two characters or less😮. I'll change it to 
`if (spaceIndex <= 0)`.

1whatleytay replied:
I saw you talk about `.split(" ")` in chat and I think that might be a more elegant solution than indexOfs and substrings.

mk5med commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+                        if (compare(origNames[i, 1], lastName) ||
+                            compare(lastName, origNames[i, 1])) {
+                            // Checking if the first names match
+                            if (compare(origNames[i, 0], firstName) ||
+                                compare(firstName, origNames[i, 0])) {
+                                // Assigning user a role specified in the command
+                                var role = Context.Guild.GetRole(roleId);
+                                await (user as IGuildUser).AddRoleAsync(role);
+                                discordMessage +=
+                                    $"`{firstName} {lastName}` has been granted role `{role}`\n";
+                                break;
```
How will this handle cases where the `lastName` is unique to the server and the `firstname` and `nickname` are not provided through Discord or not known?

muizzzafar11 replied:
I'm guessing you're talking about the case where the last name matches but the first name doesn't. I have this code in case that happens [line 87-92]
```
noLastNameMsg += $"`{lastName}` of `{firstName} {lastName}` matches " +
                                  $"with `{origNames[i, 1]}` of `{origNames[i, 0]} {origNames[i, 1]}`\n";
Console.WriteLine($"`{lastName}` of `{firstName} {lastName}` matches " +
                              $"with `{origNames[i, 1]}` of `{origNames[i, 0]} {origNames[i, 1]}`");
```

mk5med commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            int messageLimit = 2000;
+            int startIndex = 0;
+
+            if (mainMessage == null) {
+                return;
+            }
```
The brackets are not needed.

1whatleytay replied:
* but are appropriate depending on the style. 

I feel kinda uncomfortable when something like this is brought up, it makes single statement ifs with brackets sound like bad practice when it really isn't. Long statements deserve brackets. If the if statement is already within another statement with brackets (like a for) it looks much better with brackets. If/else statements can become really hard to read without brackets (and encourage falsy/broken statements where the first statement gets executed but the second isn't technically part of the if), etc.

I suppose `return` is short enough though. I will leave it up to good Muhammad man.

mk5med commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+
+        private async Task sendSplitMessage(string mainMessage, string separator) {
+            // The max limit set by discord
+            int messageLimit = 2000;
```
If this won't be changed, it should be `readonly` and defined at the beginning of the class.

1whatleytay replied:
`readonly` isn't appropriate for constant integers. Use [`const`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/const) instead.

mk5med commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            }
+
+            do {
+                // Get the substring of the specified length
+                string subString = mainMessage.Substring(startIndex, messageLimit + 1);
+                // Get the last index of new line in the substring of specified length
+                int endIndex = subString.LastIndexOf(separator);
+                // Send the message from the start index to last index of new line character
+                await ReplyAsync(mainMessage.Substring(startIndex, endIndex));
+                // set the start index to end index, +1 to skip over \n character
+                startIndex += endIndex + 1;
```
I recommend adding a blank line between comments as I found it improves readability.

mk5med commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+                return;
+            }
+
+            do {
+                // Get the substring of the specified length
+                string subString = mainMessage.Substring(startIndex, messageLimit + 1);
+                // Get the last index of new line in the substring of specified length
+                int endIndex = subString.LastIndexOf(separator);
+                // Send the message from the start index to last index of new line character
+                await ReplyAsync(mainMessage.Substring(startIndex, endIndex));
+                // set the start index to end index, +1 to skip over \n character
+                startIndex += endIndex + 1;
+            } while (startIndex + messageLimit <= mainMessage.Length);
```
If you are sending separate messages based on a singular delimiter, you can use `string#Split(Char[])` on `mainMessage` and iterate through the array and send each message.

To further expand the solution, if you are expecting that you will be creating a buffer of multiple messages, you can also use a `List<string>` and iterate through it when you are ready to send the messages. This would remove the need to creating a new array every time a message queue needs to be sent.

mk5med reviewed (changes requested):
Thanks Muhammad, this looks good! In regards to the logic of the IRL algorithm, it would be great if there was a set of automated tests that we can run to verify that the code works as expected (similar to those we have in [Passport](https://github.com/Diophontine/passport/tree/master/Passport/Tests)), and to prepare for edge-cases in advance.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+namespace TU20Bot.Commands {
+    public class NameMatch : ModuleBase<SocketCommandContext> {
+
+        private readonly string[,] origNames = {
```
This should be abstracted out to config or a function param at least.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            {"Rick","(O20)"},
+            {"Pranav","Vyas"},
+            {"Karen","Truong"},
+            {"Muhammad Muizz","nolastname"},
```
It's reasonable to use `null` here.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            {"Hargun","nolastname"},
+            {"Markos","Georghiades"},
+            {"Muhammad Muizz","Zafar"},
+            {"Taylor desgroup","Whatley"},
```
It's reasonable to say that our data may be given as a full name instead of a split first last name. Its strange how "desgroup" was packed in here. Perhaps just make it a single string a split it?

Or use match criteria that can match any of `string fullName;` or `string discordName;`

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            string noLastNameMsg = "";
+            string noFirstNameMsg = "";
+            string notInGuildMsg = "";
+            for (int i = 0; i < origNames.GetLength(0); i++) {
```
Erm...

```suggestion
            for (int i = 0; i < origNames[i].Length; i++) {
```

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            string noFirstNameMsg = "";
+            string notInGuildMsg = "";
+            for (int i = 0; i < origNames.GetLength(0); i++) {
+                foreach (var user in users) {
```
It may be more efficient if you flipped these two for loops.

I'm under the impression that Guild.Users is a computed property and this may reduce the amount of requests to discord.

```
foreach (var user in users) {
  for (int i = 0; i < origNames[i].Length; i++) {
    ...
```

muizzzafar11 replied:
But wouldn't this way be a bit slower than what we currently have?

1whatleytay replied:
Why do you think that? 🙂 

My theory is that every time we get a user object out of the users list, the value is queried from discord. If this is not true I suppose they are of equal performance but this is what experimentation is for!! :D 

Let our lists be
```
users -> user1, user2, user3
origNames -> name1, name2, name3
```

Tracing of the existing loop:
```
Getting name1 from list
Getting user1 from discord
Getting user2 from discord
Getting user3 from discord
Getting name2 from list
Getting user1 from discord
Getting user2 from discord
Getting user3 from discord
Getting name3 from list
...
```

Tracing of the suggested loop (theoretically):
```
Getting user1 from discord
Getting name1 from list
Getting name2 from list
Getting name3 from list
Getting user2 from discord
Getting name1 from list
Getting name2 from list
Getting name3 from list
Getting user3 from discord
...
```

I assume in the second on there are less requests to discord -> more performance + rate limit runs out slower?

muizzzafar11 replied:
I was looking at it from more of a total number of times the loop iterates. Since the list of users, we get from discord is bigger than the list of users we externally provide the bot with. 
```
users -> user1, user2, user3, user4, user5, user6, user7
origNames -> name1, name2
```

```
Getting name1 from list
Getting user1 from discord (name doesn't match)
Getting user2 from discord (name doesn't match)
Getting user3 from discord (name doesn't match)
Getting user4 from discord (name doesn't match)
Getting user5 from discord (name matches)

break the loop and get the next user in the list
```

```
Getting user1 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user2 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user3 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user4 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user5 from discord
Getting name1 from the list(name matches)

break the loop and get the next user from discord
```

muizzzafar11 replied:
But I'll try both of the methods and see which one is better. Also, I didn't think I would be querying discord every time I get a user in the for loop bcz I thought that I already got the users list from discord and that was the end of the "communicating with discord" part. Do you remember the message .net was giving me: `A MessageReceived handler is blocking the gateway task.` This might be the rzn why.

1whatleytay replied:
> This might be the rzn why.

Yup that's why I brought this up :D

> Since the list of users, we get from discord is bigger than the list of users we externally provide the bot with.

A couple cool theory things!~

 - Since the order of the lists are undefined, both your solution and my solution perform identically. It's a matter of chance if the first user breaks on the first iteration or last iteration (depending on how the list is ordered). Neither will outperform the other (if both lists have O(1) fetch time).
 - You can define C# properties like this that will only get the information you need. I think it's possible that every iteration of that loop it will make a new request, although I am not sure.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+                    string fullName = user.Nickname;
+                    if (fullName == null) {
+                        fullName = user.ToString();
+                        fullName = fullName.Substring(0, fullName.LastIndexOf('#'));
```
This seems to be the equivalent of `fullName = user.Username`.

```cs
fullName = user.Nickname ?? user.Username;
```

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+
+                        // Checking if the last names match
+                        if (compare(origNames[i, 1], lastName) ||
+                            compare(lastName, origNames[i, 1])) {
```
Seems appropriate to define `compare` so that it compares these parameters both ways, instead of calling it twice every time you need to use it.

1whatleytay replied:
`[i, 1]` and `[i, 0]` ring as magic numbers to me here.

Something like
```
const var firstNameIndex = 0;
const var lastNameIndex = 1;
```
might be appropriate?

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+                            if (compare(origNames[i, 0], firstName) ||
+                                compare(firstName, origNames[i, 0])) {
+                                // Assigning user a role specified in the command
+                                var role = Context.Guild.GetRole(roleId);
```
This will make a request to discord's servers. It would be smart to move this before the loop so it doesn't get requested every time.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+                                compare(firstName, origNames[i, 0])) {
+                                // Assigning user a role specified in the command
+                                var role = Context.Guild.GetRole(roleId);
+                                await (user as IGuildUser).AddRoleAsync(role);
```
Erm... Rider will yell at me for this even though its unlikely user won't be an IGuildUser. Can you use C-style cast? It will make sure a CastException is thrown instead of a NullPointer exception in the worst case scenario (world is ending).
```suggestion
                                await ((IGuildUser) user).AddRoleAsync(role);
```

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+                                var role = Context.Guild.GetRole(roleId);
+                                await (user as IGuildUser).AddRoleAsync(role);
+                                discordMessage +=
+                                    $"`{firstName} {lastName}` has been granted role `{role}`\n";
```
Being explicit about what you are trying to get from the role is probably good practice. I propose something like [`role.Name`](https://discord.foxbot.me/docs/api/Discord.IRole.html#Discord_IRole_Name).

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+                            noLastNameMsg += $"`{lastName}` of `{firstName} {lastName}` matches " +
+                                             $"with `{origNames[i, 1]}` of `{origNames[i, 0]} {origNames[i, 1]}`\n";
+                            Console.WriteLine($"`{lastName}` of `{firstName} {lastName}` matches " +
+                                              $"with `{origNames[i, 1]}` of `{origNames[i, 0]} {origNames[i, 1]}`");
```
Since this text is lengthy, it would be probably better practice to build it once and assign it to a variable, then append and WriteLine (instead of building it twice).

muizzzafar11 replied:
could you explain this with an example perhaps

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            string discordMessage = "";
+            string noLastNameMsg = "";
+            string noFirstNameMsg = "";
+            string notInGuildMsg = "";
```
Can we merge these together into `noNameMessages` or `errorLog`?

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+        [Summary("matches guild member names with a specified list and assigns roles to the matched names")]
+        public async Task matchNames(ulong roleId) {
+            var users = Context.Guild.Users;
+            string discordMessage = "";
```
For building large strings, it may be appropriate to use a [`StringBuilder`](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder?view=netcore-3.1).

Generally more performant.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,154 @@
+            discordMessage += noLastNameMsg + noFirstNameMsg;
+            // For reporting if the user doesn't exist in the server (no last name match)
+            for (int i = 0; i < origNames.GetLength(0); i++) {
+                if (!discordMessage.Contains(origNames[i, 1])) {
```
Erm... I'm not comfortable with this. It's not an efficient or safe way to accomplish this.
 - `Contains` takes a decent while, especially in large strings.
 - discordMessage can contain whatever, including user input. The algorithm could've matched `Bobby Flobby` but not `Mc Flob` but `discordMessage.Contains("Flob")` will still return true.

I suggest creating a list inside the main "matching" for loop above. Keep track of everyone who passed (or remove people from that list if they get matched). You can then just go through that list and print out the details of everyone who didn't match.

1whatleytay reviewed (changes requested):
Big review. Awesome PR :D

muizzzafar11 commented:
Thanks, Viktor and Taylor for the review. I have made most of the changes and the ones left will be done by today or tomorrow.

muizzzafar11 committed:
added minor review changes

muizzzafar11 committed:
Added unit test for name matching

muizzzafar11 committed:
inverted for loop and made message changes

muizzzafar11 committed:
added optimized solution for finding a person no in server

muizzzafar11 commented:
Aight, I have made all the changes requested in the review. Can someone look over and lmk if there's anything else to improve. Also, for the unit test, the developer will have to manually put in the token.txt file with the token in it in the current working directory.

muizzzafar11 committed:
changed names and removed unused items

muizzzafar11 committed:
initial commit email verify

muizzzafar11 committed:
made htread static for other thread to access data without instantiating everytime

muizzzafar11 committed:
email verifier fully functional, role part incomplete

muizzzafar11 commented:
Now, this PR contains an email verification system as well. Currently, the unverified emails are being saved in a `Dictionary` and are checked in a separate thread for verification. In the future, these will be stored inside a database along with the `userId`.  

The only part that's left is how to assign a role to the emails which have been verified. I'm thinking that after the frontend is implemented, we can give the bot a `roleId` in the frontend. So, whenever a user verifies his/her email, we can assign them a role based on the Id given. Is it possible to implement this system or something similar to it, @1whatleytay ? 

In addition to assigning a role to a verified user, it will also help us simplify the name match command. So, instead of `-lsuser-assignrole 750153128460025867` we can use `-lsuser-assignrole`. What do you think @viktorKorolyuk ?

mk5med committed:
Separate the IRL name matching algorithm to simplify file structure and group related logic

Also simplified compare and list checking methods

mk5med committed:
Update NameMatchTest to test the logic of the algorithm

Added an option to enable sending certain test outputs to Discord

mk5med committed:
Add the BotTest project to the TU20Bot solution

mk5med commented:
Thanks for the updates @muizzzafar11, I added a couple changes to `NameMatch.cs` to keep it simple and allow the algorithm to be tested; let me know if you have any questions or concerns with the new structure.

We discussed during a call earlier this month that the IRL will be controlled from the web console as it simplifies the process of inputting user lists for events. What we could do through Discord is allow executives to trigger a new name-based matching scan with a command similar to `<command> <event name or id>`.

I like the idea of the frontend to provide a roleID to be used by the bot. To expand further, the bot should ideally be able to add roles to the user based on multiple "event" attendee lists; this would be used in cases to handle lists of mentors, membership holders, or for individuals who participate in multiple of our initiatives or events.

1whatleytay commented:
Hey Viktor,

We have an existing style for the rest of the project. It might be worth looking over the projects existing code and modifying your commits to comply 🙂

1whatleytay commented on BotTest/EmailVerificationTest.cs:
```diff
@@ -0,0 +1,87 @@
+            // Start bot with token from "token.txt" in working folder.
+            try {
+                var _token = File.ReadAllText("token.txt").Trim();
+                token = _token.ToString();
```
`_token` is a `String` already.

```suggestion
                token = File.ReadAllText("token.txt");
```

1whatleytay commented on BotTest/EmailVerificationTest.cs:
```diff
@@ -0,0 +1,87 @@
+            // Initializing all the requored classes
+            _config = new Config();
+            _client = new Client(_config);
+            _handler = new Handler(_client);
```
You actually don't need to initialize Handler here I don't think. Handler sets up events and starts up the command API. Since tests won't be using that we can drop it I think.

1whatleytay commented on TU20Bot/Commands/EmailVerification.cs:
```diff
@@ -0,0 +1,36 @@
+namespace TU20Bot.Commands {
+    public class EmailVerification : ModuleBase<SocketCommandContext> {
+
+        private Config config = new Config();
```
Erm... This wasn't so clear but you can do `((Client)Context.Client).config` to access the config object. Creating a new one will cause problems.

1whatleytay commented on TU20Bot/Commands/EmailVerification.cs:
```diff
@@ -0,0 +1,36 @@
+                await ReplyAsync("Could not verify email. Your email has been saved and will be verified automatically.");
+        }
+
+        public string emailCompare(string email, ulong userId) {
```
I think its appropriate to make this `static`. Seems like it would simplify testing code.

1whatleytay commented on TU20Bot/Commands/EmailVerification.cs:
```diff
@@ -0,0 +1,36 @@
+            // Since the email didn't match, adding user to the dictionary
+            Config.userEmailId.Add(userId, email);
+            return null;
+        }
```
Note how this function always returns either `"Email verified"` or `null`. There are some ways we can improve this.
 - If a function can have more than 2 outputs, we tend to use an [enum](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/enum). I don't think this is nessesary but it is better than returning a String (enum compares are stricter, type checked by the language and faster).
 - We could make it [throw](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/throw) an error if the email doesn't match. This also has the benefit of type checking and faster performance but its still a bit overkill for 2 outputs.

So I suggest why don't you just return a `bool` ? Return true when it finds an email and false when it doesn't?

1whatleytay commented on TU20Bot/Commands/EmailVerification.cs:
```diff
@@ -0,0 +1,36 @@
+            for (int i = 0; i < config.emails.Count; i++) {
+                if (email.Equals(config.emails[i]))
+                    return "Email verified";
+            }
```
I believe you can make this a linq query :O It's a bit shorter than a for loop.

`bool emailIsInList = config.emails.Any(x => x == email);`

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,242 @@
+            await sendSplitMessage(await nameMatching(_config.origNames, users, roleId), "\n");
+        }
+
+        public async Task<string> nameMatching(string[,] listNames, IReadOnlyCollection<SocketGuildUser> users, ulong? roleId)
```
Sorry, but now looking at [line 19](https://github.com/TechUnderTwenty/TU20Bot/pull/4/files#diff-3b53b44614038416e8921c727c2d464fR19) doesn't it make more sense for listNames to be a list of structs?

```
class ListName {
  string first;
  string last;
}

...

ListName[] listNames;

foreach (name in listNames) {
  if (compare(name.first)) { ... }
  if (compare(name.last)) { ... }
}
```

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,242 @@
+        {
+
+            StringBuilder errorLog = new StringBuilder("");
+            StringBuilder discordMessage = new StringBuilder("");
```
```suggestion
            var errorLog = new StringBuilder();
            var discordMessage = new StringBuilder();
```

Nit pick here really. Feel free to use `var` when you already declare the type of the variable in the expression. Saves you some typing time and looks a bit nicer imo.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,242 @@
+        private const int firstNameIndex = 0;
+        private const int lastNameIndex = 1;
+
+        public delegate void IndexMatch(int index);
+        public delegate void IndexMatchDetailed(int index, string firstName, string lastName);
```
Delegates have some overhead that we aren't using here. `Func` might be more appropriate.

mk5med replied:
> Delegates have some overhead that we aren't using here. Func might be more appropriate.

```csharp
public delegate TResult Func<in T,out TResult>(T arg);
public delegate void IndexMatch(int index);
```
[Source](https://docs.microsoft.com/en-us/dotnet/api/system.func-1?view=netcore-3.1)

This sounds interesting, would you mind elaborating? In C# `Func` is a delegate and a delegate defines the data type for the lambda, which would mean that using a custom delegate and a `Func` within the method are functionally equivalent; the only difference is that a `Func` is already defined within the System namespace, and that a Func requires a return value.

In the case for how these delegates were used, using a `Func` or `Action` would mean defining the argument types (and, in the case of `Func`, the return type) within the `nameNatchAlg` method definition itself, or by defining a constant set of listeners and using those. 

In the former case, this concentrates the definition to a localised area of the file, which is not ideal as it becomes very verbose and code is repeated, which reduces maintainability.

In the latter case, access to the contextual variables is lost (such as the message lists), which doesn't automatically work with the Discord-side of the algorithm design as it was, and it would have made the testing more complex.

1whatleytay replied:
Hmm, I'm not sure about that.

Delegate defines the `+=` operator and works with events while `Func<>` seems to just act aking to a C function pointer.\

You don't need to define it inside the function method itself if you use `using x = y`. It's how C# aliases types iirc.

mk5med replied:
Would you mind elaborating more? A [Func is a delegate](https://docs.microsoft.com/en-us/dotnet/api/system.func-1?view=netcore-3.1), which means that it would also have the `+=` operator.
`Func<int>` would be defined as: `public delegate TResult Func<out TResult>();`.

In the case of using `using MethodToCall = Func<int, int>;` or `using MethodToCall = Action<int, int>;`:
```csharp
// At the top of the namespace block
using MethodToCall = Action<int, int>; 
// -> delegate void System.Action<in T1, in T2>(T1 arg1, T2 arg2)

// Is functionally equivalent to
delegate void MethodToCall(int index, int otherValue); 
// -> delegate void TU20Bot.Commands.NameMatch.MethodToCall(int index, int otherValue)
```

Both are functionally equivalent, however they are different types.

1whatleytay replied:
Hmm, that's my bad. However, I'm confident there's a way to get a raw function pointer in C#. Func just might not be it.

Use either one then :)

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,242 @@
+
+        // The max limit of chars set by discord
+        private const int messageLimit = 2000;
+        private Config _config;
```
It's worth using `((Client)Context.Client).config` here as well 🙂

1whatleytay commented on BotTest/NameMatchTest.cs:
```diff
@@ -0,0 +1,148 @@
+            // Start bot with token from "token.txt" in working folder.
+            try
+            {
+                var _token = File.ReadAllText("../../../../token.txt").Trim();
```
`"../../../../"` seems like it was left here by mistake?

mk5med replied:
This relative reference is so that we don't need to add a duplicate or link of `token.txt` into the `BotTest` project. This line will need to be tested in a Windows environment as they use a backslash separator for paths.

1whatleytay replied:
Windows can use either `/` or `\`, no need to worry.

I don't really see how this is effective. If you want to specify a location to store both token.txts you change the current directory in the project settings. This is workaroundy and takes away the ability to do the above.

mk5med replied:
> Windows can use either `/` or `\`

That's great! I have also confirmed this on my test system.

In terms of efficiency, this solution isn't great as it adds an extra maintenance step if the test output directory ever changes. Thanks for bringing notice to this line, I found a simple solution where we can automatically link `token.txt` into the project directory by specifying the relationship within the `csproj` file.

1whatleytay commented on BotTest/NameMatchTest.cs:
```diff
@@ -0,0 +1,148 @@
+        {
+
+            string token = "";
+            testContext.WriteLine("Iteration 1");
```
This perhaps was left here by mistake as well.

1whatleytay commented on TU20Bot/Configuration/Config.cs:
```diff
@@ -30,7 +30,46 @@ public class Config {
+            "johndoe@nothingmail.com",
+        };
+
+        public static Dictionary<ulong, string> userEmailId = new Dictionary<ulong, string>();
```
Know that I did mean for these to be not `static`. If they're static they wont be serialized :)

muizzzafar11 replied:
But if they're static, every time I instantiate the `Config` class by using `new Config()`, the data stored in the `userEmailId` is lost, since it's initialized again :(

Do you think it would be a good idea to instantiate the Config class only once throughout the whole program (at the start of the program, in the `Program.cs` file), `userEmailId` will retain its data? 🤔 That way I won't have to make it static.

1whatleytay replied:
You should never instantiate Config. :/

We *have* on instance shared across the program. This is explained in other messages. If you can't find the instance in a certain place you can ask me.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,242 @@
+        /// <param name="firstOrLastMatch">Executed when there is no space in fullname, however either the first or last name match</param>
+        /// <param name="succesfulMatch">Executed when there is a complete match, that is both first name and last name match</param>
+        /// <param name="lastNameMatch">Executed when only the last name matches </param>
+        public static void nameMatchAlg(string fullName, string[,] listNames, IndexMatch firstOrLastMatch, IndexMatchDetailed succesfulMatch, IndexMatchDetailed lastNameMatch)
```
Sorry @viktorKorolyuk this delegate system might need a bit more explaining to me. They seem like they are working around a return value, which would generally give more control over to the parent function and potentially increase the usefulness of the algorithm.

I suggest:
```
enum NameMatchLevel {
  // NoMatch ?
  NoSpacesMatch,
  LastNameMatch,
  CompleteMatch
}

class NameMatchProfile {
  NameMatchLevel level;
  string fullName;
}

public static List<NameMatchProfile> matchNames(string fullName, ListName[] names) { ... }
```

The additional control can come in the form of queries to the list. For example:
```
var matches = matchNames(...);

matches.IsEmpty() // no matches

matches.FirstOrDefault(x => x.level == NameMatchLevel.CompleteMatch) // find the first complete match
matches.Count(x => x.level == NameMatchLevel.CompleteMatch) // how many people fully matched?

matches.Count(x => x.name.StartsWith("John")) // analyze names
```

Generally, this can also be done with delegates, but its more complicated. Variables need to be used to track specfic information. And it seems like the tests pass empty lambdas just so this function can be called.

mk5med replied:
I like that idea! If we create a list from the parsed data the file structure will be flatter and the related logic can be grouped together even more.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,242 @@
+            // For last name since last name should be exactly the same
+            if (fullMatch)
+            {
+                return name.Replace(" ", "").Equals(matchName.Replace(" ", "")) || matchName.Replace(" ", "").Equals(name.Replace(" ", ""));
```
Generally `a == b` is the same as `b == a`.

```suggestion
                return name.Replace(" ", "") == matchName.Replace(" ", "");
```

1whatleytay commented on TU20Bot/Program.cs:
```diff
@@ -25,16 +26,24 @@ internal class Program {
             // Run server on another thread.
             new Thread(() => server.RunAsync().GetAwaiter().GetResult()).Start();
 
+            // Run email checker on another thread
+            new Thread(new ThreadStart(() => {
+                // Keep the thread running forever
+                do {
+                    new EmailChecker(config, client).checkForEmail();
+                } while (true);
+            })).Start();
```
A thread is not nessesary for this... It will take up a fair bit of computing power, it perhaps would make sense if you had run the check every 1 minute or something (via Thread.Sleep)?

I think you can tie this to some sort of event instead of a forever running thread. For example
```
someUserEmailChangingFunction() {
  // change something to do with emails...
  EmailChecker.checkForEmail(); // run the check!!
}

// Maybe you need to run it every time something happens in discord, DiscordSocketClient has your back!
client.UserJoined += () => {
  EmailChecker.checkForEmails(); // run every time someone joins the server
}

client.MessageRecieved = () => {
  EmailChecker.checkForEmails(); // run every time someone sends a message
}

client.RoleAssigned = () => {
  EmailChecker.checkForEmails(); // run every time some gets assigned a role
}
```

P.S. if you use the discord methods you will usually have to take in some parameter (usually some kind of User/Message/Role/Guild object) and probably return `Task.CompletedTask`.

muizzzafar11 replied:
That's a good idea. I'm thinking of just running it once every hour, instead of tying it to an event. Cuz in case that event doesn't occur for a certain period of time (say five or six hours), we will have a delay in assigning the roles.

muizzzafar11 replied:
Its also cuz most of the discord events are based on human activity and I don't trust those inferior creatures known as humans. 👽

1whatleytay replied:
:/ I don't know what requires something like this though. If an email can only be added every time someone sends a message or joins the server etc. you should run this check on that event. That way:
 - [x] super efficient
 - [x] no delay

1whatleytay reviewed (changes requested):
Some nit picks. Feel free to explain anything if I misunderstood :D

mk5med committed:
Update TU20Bot/Commands/NameMatch.cs

Co-authored-by: 1whatleytay <1whatleytay@hdsb.ca>

mk5med committed:
Add a custom style for C# code format rules

mk5med committed:
Remove use of callbacks to simplify and flatten name matching algorithm

And reformat code to follow project style

mk5med commented:
> We have an existing style for the rest of the project. It might be worth looking over the projects existing code and modifying your commits to comply 🙂

Sounds good, I added the style into the solution's code format rules.

mk5med committed:
Add a link to token.txt in the BotTest folder

muizzzafar11 committed:
implemented review changes

muizzzafar11 committed:
Merge branch 'IRL' of https://github.com/TechUnderTwenty/TU20Bot into IRL

muizzzafar11 committed:
fixed accidental commit

muizzzafar11 committed:
Updated nameMatchTest.cs accidental commit

muizzzafar11 committed:
Update BotTest.csproj

muizzzafar11 committed:
Updated NameMatchTest.cs

muizzzafar11 committed:
Implemented csv reader, changes to test need to be made

muizzzafar11 committed:
update .csproj file

muizzzafar11 committed:
changes to tests and methods

Two tests are still failing (problem with the name mathcing algorithm)

muizzzafar11 committed:
fixed namematch test input for nameMatchAlg

mk5med commented on TU20Bot/Properties/launchSettings.json:
```diff
@@ -0,0 +1,8 @@
+  "profiles": {
+    "TU20Bot": {
+      "commandName": "Project",
+      "commandLineArgs": "init NzM2NzQ2Mzg0NzIzNjczMTY5.XxzSig.rn1vRN_RlKCR3wzhzrASYCoFuGA"
```
It's important to remember to protect secret keys as they can be used by others to take arbitrary control over the functions of the application. No worries in this case, I'll be sending a new token for everyone!

muizzzafar11 replied:
Whoops. Sorry, I forgot to exclude it from the changes staged for commit.

mk5med commented on TU20Bot/Commands/EmailVerification.cs:
```diff
@@ -0,0 +1,44 @@
+                if (email.Equals(userEmail.Email))
+                    return true;
+            }
+            return false;
```
You can also use list predicates to shorten this check.

```suggestion
            return csvEmailList.Exists(e => e.Email == email);
```

mk5med committed:
Add SharedVolume directory and trim element output from CsvReader

mk5med commented on TU20Bot/Configuration/EmailChecker.cs:
```diff
@@ -0,0 +1,30 @@
+            this.client = client;
+        }
+
+        public void checkForEmail(List<CSVData> csvEmail) {
+            for (int i = 0; i < config.userEmailId.Count; i++) {
+                if (csvEmail.Any(x => x.Email.Equals(config.userEmailId.ElementAt(i).Value))) {
+                    ulong userId = config.userEmailId.ElementAt(i).Key;
+                    var user = client.GetUser(userId);
+                    Console.WriteLine($"{user} email verified from list");
+                    // Remove that specific index from the dictionary since the user has been verified
+                    config.userEmailId.Remove(userId);
+                }
+            }            
```
Adding comments and documentation for what this method does would be ideal here.

mk5med reviewed (approved):
Thanks for your work @muizzzafar11, although there are some minor updates for the code, based on my tests, all the new features appear to be working!

muizzzafar11 committed:
review suggestiong implemented and removed extra code form test

1whatleytay commented on BotTest/NameMatchTest.cs:
```diff
@@ -0,0 +1,120 @@
+            // Start bot with token from "token.txt" in working folder.
+            try {
+                var _token = File.ReadAllText("token.txt").Trim();
+                token = _token.ToString();
```
here :)

1whatleytay commented on BotTest/NameMatchTest.cs:
```diff
@@ -0,0 +1,120 @@
+        [TestMethod]
+        public void TestMatchNameAlgorithm_LastNameMatch() {
+            List<CSVData> nameSet = new List<CSVData>{
+            new CSVData { FirstName = "Bill", LastName = "Doe", Email = "johndoe@tu20.com" },
```
```suggestion
                new CSVData { FirstName = "Bill", LastName = "Doe", Email = "johndoe@tu20.com" },
```

1whatleytay replied:
This object is created 3 times... moving it to a static variable can reduce code size.

1whatleytay commented on TU20Bot/Program.cs:
```diff
@@ -24,17 +26,32 @@ internal class Program {
             // Run server on another thread.
             new Thread(() => server.RunAsync().GetAwaiter().GetResult()).Start();
 
+            // Run email checker on another thread every 15 min
+            new Thread(new ThreadStart(() => {
+
+                do {
+                    // Reading and assigning data from csv file
+                    config.userDataCsv = csvReader.readFile();
+
+                    new EmailChecker(config, client).checkForEmail(config.userDataCsv);
+                    Thread.Sleep(1800000);
+
+                } while (true);
+            })).Start();
```
I don't feel comfortable approving while this is here...

muizzzafar11 committed:
role asigning based on attendee or speaker, fixes from review

muizzzafar11 commented:
With the last commit, the bot can detect whether a person is an attendee or a speaker and assign them roles automatically. The algorithms I've used aren't optimized and there's a lot of room for improvement, like for starters, using enums for role detection instead of bool. I'll implement those changes and push the code sometime soon. I just wanted to get it ready for the expo. Also, the tests are failing. I'll push the code for the test tomorrow (today) or the day after tomorrow (tomorrow). Hoe the tests failing aren't a problem since I tested it in the channel and it appears to be working fine.

muizzzafar11 committed:
Fixed Unit test error, made code more modular, minor bug fixes

muizzzafar11 committed:
initial commit

muizzzafar11 committed:
initial commit -2

muizzzafar11 committed:
added basic comm b/w db and bot, can verify email from db and assign userrole

muizzzafar11 committed:
fixed bug and replaced parameters for test func

muizzzafar11 committed:
removed extra code from config

muizzzafar11 commented:
With the last commit, the bot is now able to store unverified emails inside a table in the DB (PostgreSQL). Now, the user (person managing the bot) just has to edit data in the csv file, the bot will handle the rest. To utilize the db features, make sure to change the password to your passportdev user password inside the connection string and run `Add-Migration InitialCreate` and `Update-Database` inside the package manager.  More detail on how to create migrations [here](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/?tabs=vs).

I am looking for suggestions to improve the code quality/readibility. One thing I was thinking of was to store the postgres connection string inside a `.json` file, similar to what we have in passport. Do you think I put the connection string in a new `.json` file for it or store it in the `launchsettings.json`? If you have any more suggestions regarding the placement of files or how the code is being managed please let me know :)

mk5med committed:
Merge branch 'master' into IRL

mk5med committed:
Update project for containerisation

docker-compose.yml now starts postgres db and frontend
BotDbContext.cs now changes the db reference based on environmental variable
Server.cs now listens to all ports on the system

mk5med committed:
Add container and compose file for administration

mk5med committed:
Update EmailVerification to allow verification of roles through direct messages

mk5med committed:
Add InitialCreate migration to the project

mk5med committed:
Update EmailVerification to remove last message

mk5med committed:
Restructure TU20Bot project folder hierarchy

mk5med commented:
Thanks Muhammad, this looks great! I added some changes to the PR to allow for containerisation of the project and to support email-verification of users through direct-message channels. Currently the password is hardcoded for quick deployment onto the server.

Having an `appsettings.json` file would be helpful for the database credential configuration and we will be able to work on that aspect of the automated project setup scripts in the future.

1whatleytay commented on TU20Bot/Migrations/BotDbContextModelSnapshot.cs:
```diff
@@ -0,0 +1,41 @@
+    {
+        protected override void BuildModel(ModelBuilder modelBuilder)
+        {
+#pragma warning disable 612, 618
```
??

1whatleytay replied:
Uhh... using pragma to disable warnings is a bit strange. Seems like it might hide errors that could be helpful.

1whatleytay commented on TU20Bot/Migrations/BotDbContextModelSnapshot.cs:
```diff
@@ -0,0 +1,41 @@
+
+            modelBuilder.Entity("TU20Bot.Configuration.UnverifiedUser", b =>
+                {
+                    b.Property<int>("Id")
+                        .ValueGeneratedOnAdd()
+                        .HasColumnType("integer")
+                        .HasAnnotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.IdentityByDefaultColumn);
+
+                    b.Property<string>("Email")
+                        .HasColumnType("text");
+
+                    b.Property<decimal>("UserId")
+                        .HasColumnType("numeric(20,0)");
+
+                    b.HasKey("Id");
+
+                    b.ToTable("unverifiedUsers");
```
Is this in place of decorations.

1whatleytay commented on TU20Bot/Migrations/20200919214442_InitialCreate.Designer.cs:
```diff
@@ -0,0 +1,43 @@
+    {
+        protected override void BuildTargetModel(ModelBuilder modelBuilder)
+        {
+#pragma warning disable 612, 618
```
Erm...

1whatleytay commented on TU20Bot/Migrations/20200919214442_InitialCreate.Designer.cs:
```diff
@@ -0,0 +1,43 @@
+
+            modelBuilder.Entity("TU20Bot.Configuration.UnverifiedUser", b =>
+                {
+                    b.Property<int>("Id")
+                        .ValueGeneratedOnAdd()
+                        .HasColumnType("integer")
+                        .HasAnnotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.IdentityByDefaultColumn);
+
+                    b.Property<string>("Email")
+                        .HasColumnType("text");
+
+                    b.Property<decimal>("UserId")
+                        .HasColumnType("numeric(20,0)");
+
+                    b.HasKey("Id");
+
+                    b.ToTable("unverifiedUsers");
```
Decorations again?

1whatleytay commented on TU20Bot/Database/BotDbContext.cs:
```diff
@@ -0,0 +1,16 @@
+
+        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) {
+            var server = Environment.GetEnvironmentVariable("ISCONTAINER") == "TRUE" ? "db" : "127.0.0.1";
+            optionsBuilder.UseNpgsql($"Server={server};Port=5432;Database=discordbot;User Id=passportdev;Password=test;");
```
Oh yeah also we can't have this. It exposes our UN/PW, which could be what's hosting on our DO right now :flushed:

Some configuration file/default is appropriate.

1whatleytay commented:
<img width="987" alt="Screen Shot 2020-09-24 at 9 46 16 AM" src="https://user-images.githubusercontent.com/32211852/94153515-e0286480-fe4a-11ea-87b9-219c60df2dc0.png">
I think it's smart to leave this out of the CS project file and move them over to the dockerfile. The .csv doesn't really have anything to do with the core project anyway.

We also have a way to create token.txt via command line arguments.

1whatleytay commented on TU20Bot/Program.cs:
```diff
@@ -24,17 +27,33 @@ internal class Program {
             new Thread(() => server.RunAsync().GetAwaiter().GetResult()).Start();
 
+            // Run email checker on another thread every 15 min
+            new Thread(new ThreadStart(async () => {
+
+                do {
+                    // Reading and assigning data from csv file every 15 min
+                    config.userDataCsv = csvReader.readFile();
+
+                    await new EmailChecker(config, client, dbCommUnverifiedUser).emailCheck(config.userDataCsv);
+                    await Task.Delay(1800000);
+
+                } while (true);
+            })).Start();
```
I cannot merge this with this here.

Please move this to any other event. If you need it to run every time the csv file changes, use [FileWatcher](https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher?view=netcore-3.1). If you need it to run every time the user runs a verify command, call it in the verify command.

This is too much.

1whatleytay commented on TU20Bot/Models/CSVData.cs:
```diff
@@ -0,0 +1,18 @@
+﻿using CsvHelper.Configuration.Attributes;
+
+namespace TU20Bot.Configuration {
+
```
This is personal preference but... I'd rather you don't put a space at the beginning of every class and method. I don't think it serves a purpose, and the rest of the project doesn't have this. It should be tightly packed:

```
class MyClass {
  int a;
  int b;

  void x() {

  } // a nice waterfall of } at the end here too
}
```

1whatleytay commented on TU20Bot/Models/CSVData.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace TU20Bot.Configuration {
+
+    public class CSVData {
```
Name is too broad here. It sounds like this class is a general purpose for any and every csv file, but it isn't. Maybe `UserDetails`.

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,220 @@
+
+        public async Task<string> nameMatching(List<CSVData> csvlistNames, IReadOnlyCollection<SocketGuildUser> users, ulong? roleId) {
+
+            var errorLog = new StringBuilder("");
+            var discordMessage = new StringBuilder("");
```
```suggestion
            var errorLog = new StringBuilder();
            var discordMessage = new StringBuilder();
```

1whatleytay commented on TU20Bot/Commands/NameMatch.cs:
```diff
@@ -0,0 +1,220 @@
+            public string fullName { get; set; }
+        }
+
+        [Command("lsuser-assignrole")]
+        [Summary("matches guild member names with a specified list and assigns roles to the matched names")]
+        public async Task matchNames(ulong roleId) {
+
+            var users = Context.Guild.Users;
+
+            _config = ((Client)Context.Client).config;
+
+            await sendSplitMessage(await nameMatching(_config.userDataCsv, users, roleId), "\n");
+        }
```
Is this only for debugging? Can we remove it?

One thing I am worried about is regular users running commands they should probably not be able to access.

mk5med replied:
Great point! Unlike the email verification system, this shouldn't be a feature that is triggered through discord. Let's disable the command for now and we will work on triggering it from the frontend later.

1whatleytay commented on TU20Bot/Configuration/EmailChecker.cs:
```diff
@@ -0,0 +1,79 @@
+        }
+
+
+        public async Task assignRole((ulong userId, CSVData userData) userInfo) {
```
```suggestion
        public async Task assignRole(ulong userId, CSVData userData) {
```

tuplesepction whaduheck

1whatleytay committed:
Add Match Endpoints for IRL (#7)

* Cleanup IRL before matching endpoints

* Add name matching server endpoints

1whatleytay committed:
Config changes, pulling in MongoDB and starting the transition (#8)

* Cleanup IRL before match endpoints

* Add name matching server endpoints

* Refactor IRL data structure before MongoDB

* Store verified emails in MongoDB

* Update docker-compose.yml for new mongodb environment

Co-authored-by: Viktor Korolyuk <8334048+viktorKorolyuk@users.noreply.github.com>

mk5med committed:
Update docker-compose.yml

mk5med commented:
@muizzzafar11 and @1whatleytay, thanks for the great work you put in, the IRL system has been officially implemented into the bot!
